### PR TITLE
[12.1.x] ISPN-12941 Add asynchronous touch option for max-idle 

### PR DESCRIPTION
**Pull request by danberindei**
https://issues.redhat.com/browse/JDG-4690

Backport of #9258

The only production code change vs. main is to allow `touch="ASYNC"` since 12.1 instead of 13.0.
In the tests I moved the `touch="ASYNC"` configs to 12.1.xml and I included the fix for ISPN-13161.

### CODE REVIEW

**Review by wburns — APPROVED** [review]
Changes look good as far as I can see.

However we need to change master for the `CacheParser` so a user who upgrades won't get an error about async touch not being part of 12.1 schema version.

**Comment by danberindei** [comment]
Rebased. Since it's so late in the 12.1/DG 8.2 lifecycle I'm not sure it's worth changing the "since" check in main, but we can certainly do it if/when this gets integrated.

File: core/src/main/java/org/infinispan/util/logging/Log.java [hunk 1]
```diff
@@ -2192,4 +2192,7 @@ CacheConfigurationException storeConfiguredHasBothReadAndWriteOnly(String storeC
 
    @Message(value = "Failed to recover '%s' persisted data.", id = 644)
    PersistenceException corruptDataMigrationFailed(String cacheName, @Cause Throwable cause);
+
+   @Message(value = "Async cache modes like %s do not allow expiration with SYNC touch mode", id = 645)
```
Comment (oraNod) [inline core/src/main/java/org/infinispan/util/logging/Log.java:2196]:
```suggestion
   @Message(value = "Asynchronous cache modes, such as %s, cannot use SYNC touch mode for maximum idle expiration.", id = 645)
```

File: core/src/main/resources/schema/infinispan-config-12.1.xsd [hunk 2]
```diff
@@ -1203,6 +1203,15 @@
         </xs:documentation>
       </xs:annotation>
     </xs:attribute>
+    <xs:attribute name="touch" type="tns:touch-mode" default="${Expiration.touch}">
+      <xs:annotation>
+        <xs:documentation>
+          Control how the timestamp of read entries with max-idle are updated on all the key owners in a cluster.
+          Default is SYNC.
+          If the cache mode is ASYNC this attribute is ignored, and timestamps are updated asynchronously.
```
Comment (oraNod) [inline core/src/main/resources/schema/infinispan-config-12.1.xsd:1211]:
```suggestion
          Controls how timestamps get updated for entries in clustered caches with maximum idle expiration.
          The default value is SYNC.
          This attribute applies only to caches that use synchronous mode.
          Timestamps are updated asynchronously for caches that use asynchronous mode.
```

File: core/src/main/resources/schema/infinispan-config-12.1.xsd [hunk 3]
```diff
@@ -2875,6 +2884,28 @@
     </xs:restriction>
   </xs:simpleType>
 
+  <xs:simpleType name="touch-mode">
+    <xs:restriction base="xs:token">
+      <xs:enumeration value="SYNC">
+        <xs:annotation>
+          <xs:documentation>Delay read operations until the other owners confirm updating the timestamp of the
+            entry.
```
Comment (oraNod) [inline core/src/main/resources/schema/infinispan-config-12.1.xsd:2892]:
```suggestion
          <xs:documentation>Delays read operations until the other owners confirm that the timestamp for the entry is updated.
```

File: core/src/main/resources/schema/infinispan-config-12.1.xsd [hunk 4]
```diff
@@ -2875,6 +2884,28 @@
...
+          </xs:documentation>
+        </xs:annotation>
+      </xs:enumeration>
+      <xs:enumeration value="ASYNC">
+        <xs:annotation>
+          <xs:documentation>
+            Send touch commands to other owners, but do not wait for their confirmation.
+            This allows read operations to return the value of a key even if another node has started
+            expiring it.
+            When that happens, the read won't extend the lifespan of the key.
```
Comment (oraNod) [inline core/src/main/resources/schema/infinispan-config-12.1.xsd:2902]:
```suggestion
            Sends touch commands to other owners without waiting for confirmation.
            This mode allows read operations to return the value of a key even if another node has started
            expiring it.
            When that occurs, the read operation does not extend the lifespan of the key.
```

**Review by oraNod — COMMENTED** [review]
@danberindei Hey a couple of doc suggestions. Feel free to take them or leave them. I do realize this is a backport so doc edits need to go on main as well. Maybe I can add them to a PR with some other doc updates. 

One style consistency thing to note. Please stick with third person singular form for schema descriptions. "Specifies, Defines, Sets, Controls, etc". Cheers!

**Review by wburns — APPROVED** [review]
LGTM

There is a random failure though https://ci.infinispan.org/job/Infinispan/job/PR-9437/4/testReport/junit/org.infinispan.expiration.impl/ClusterExpirationMaxIdleTest/testMaxIdleReadNodeDiesPrimary_REPL_SYNC__tx_false__OFF_HEAP__touch_ASYNC_/

**Comment by wburns** [comment]
Integrated into 12.1.x, thanks @danberindei !

**Pull request merged**